src/k2/cpo/service/CPOService.cpp

```diff
@@ -183,6 +195,184 @@ CPOService::handleGet(dto::CollectionGetRequest&& request) {
+            (dto::Verbs::K23SI_PUSH_SCHEMA, request, *endpoint, 1s));
+    }
+
+    return when_all_succeed(pushFutures.begin(), pushFutures.end())
+    .then([] (auto&& doneFutures) {
+        for (const auto& doneFuture : doneFutures) {
+            auto& [status, k2response] = doneFuture;
+            if (!status.is2xxOK()) {
+                K2WARN("Failed to push schema update: " << status);
+                return seastar::make_ready_future<Status>(status);
+            }
+        }
+
+        return seastar::make_ready_future<Status>(Statuses::S200_OK(""));
+    })
+    .handle_exception([](auto exc) {
+        K2ERROR_EXC("Failed to push schema update: ", exc);
+        return seastar::make_ready_future<Status>(Statuses::S500_Internal_Server_Error("Failed to push schema"));
+    });
+}
+
+seastar::future<std::tuple<Status, dto::GetSchemasResponse>>
+CPOService::handleSchemasGet(dto::GetSchemasRequest&& request) {
+    auto it = schemas.find(request.collectionName);
+    if (it == schemas.end()) {
+        std::tuple<Status, dto::Collection> getResponse = _getCollection(request.collectionName);
+        if (!std::get<0>(getResponse).is2xxOK()) {
+            return RPCResponse(std::move(std::get<0>(getResponse)), dto::GetSchemasResponse{});
+        }
+
+        it = schemas.find(request.collectionName);
+        K2ASSERT(it != schemas.end(), "Schemas iterator is end after collection refresh");
+    }
+
+    return RPCResponse(Statuses::S200_OK("SchemasGet success"), dto::GetSchemasResponse { it->second });
+}
+
+seastar::future<std::tuple<Status, dto::CreateSchemaResponse>>
+CPOService::handleCreateSchema(dto::CreateSchemaRequest&& request) {
+    K2INFO("Received schema create request for " << request.collectionName << " : " << request.schema.name);
+
+    // 1. Stateless validation of request
+
+    Status validation = basicSchemaValidation(request.schema);
+    if (!validation.is2xxOK()) {
+        return RPCResponse(std::move(validation), dto::CreateSchemaResponse{});
+    }
+
+    // 2. Stateful validation
+
+    // getCollection to make sure in memory cache of schemas is up to date, and we need the 
+    // collection with partition map anyway to do the schema push
+    std::tuple<Status, dto::Collection> getResponse = _getCollection(request.collectionName);
+    if (!std::get<0>(getResponse).is2xxOK()) {
+        return RPCResponse(std::move(std::get<0>(getResponse)), dto::CreateSchemaResponse{});
+    }
+    dto::Collection& collection = std::get<1>(getResponse);
+
+    bool validatedKeys = false;
+    for (const dto::Schema& otherSchema : schemas[request.collectionName]) {
+        if (otherSchema.name == request.schema.name && otherSchema.version == request.schema.version) {
+            return RPCResponse(Statuses::S403_Forbidden("Schema name and version already exist"), dto::CreateSchemaResponse{});
+        }
+
+        // By induction, as long as we validate that the key fields match with at least one 
```

Comment:
I think it's a bit confusing to say that this is true by induction. i associate this with recursion, and i'd expect to see comparison with the latest version to have a proof by induction. I think i can convince myself that compatibility with any version implies compatibility of the entire chain, but in my head, that's because the only mutating operations allowed are add/rename/drop column[s], and all of these are idempotent. so maybe just a comment thing :)